modules/cache/jvm/src/test/scala-2.12/coursier/cache/FileCacheTests.scala

```diff
@@ -926,6 +926,69 @@ object FileCacheTests extends TestSuite {
       }
     }
 
+    test("custom protocols") {
+      test("unknown") - async {
+        val artifact = Artifact("unknown.protocol://hostname/file.txt")
+
+        val res = await {
+          FileCache()
+            .file(artifact)
+            .run
+            .future
+        }
+
+        val expectedReason =
+          List(
+            "Caught java.net.MalformedURLException (unknown protocol: unknown.protocol) while downloading unknown.protocol://hostname/file.txt.",
+            "Visit https://get-coursier.io/docs/extra.html#extra-protocols to learn how to handle custom protocols."
+          ).mkString(" ")
+
+        res match {
+          case Left(error: ArtifactError.DownloadError) =>
+            println("---")
+            println(error.reason)
+            println()
+            println(expectedReason)
+            println("---")
+            assert(error.reason == expectedReason)
+
+          case _ =>
+            assert(false)
+        }
+      }
+
+      test("with classloader") {
+        withTmpDir { dir =>
+          async {
+            val classloader = 
+              new URLClassLoader(
+                CustomLoaderClasspath.files.map(new URL(_)).toArray
+              )
+
+            val artifact = Artifact("customprotocol://hostname/README.md")
+
+            val res = await {
+              FileCache()
+                .withClassLoaders(Seq(classloader))
+                .withLocation(dir.toFile)
+                .file(artifact)
+                .run
+                .future
+            }
+
+            res match {
+              case Right(file: File) =>
```

Comment:
No need of the type ascription, it could make that code still compile if we were to change the return type of `FileCache.file`, which we don't want:
```suggestion
              case Right(file) =>
```